### PR TITLE
Qualcomm AI Engine Direct - Suport batch prefill mode for llama3.2

### PR DESCRIPTION
**Pull request by chunit-quic**
- Enable bert mode
- Change input sequence of static_llama
- Tag bert output as uint8
- Unify both 1b and 3b in 1 runner
- Add hybrid IO memory for llama3_2 runner
- Align timer with llama

### CODE REVIEW

**Comment by cccclai** [comment]
Hey do you mind sharing the command for AoT and runtime so I can try on my end?

**Comment by chunit-quic** [comment]
> Hey do you mind sharing the command for AoT and runtime so I can try on my end?


Sure!  Change different by set up the argument `model_mode`.
python examples/qualcomm/oss_scripts/llama3_2/llama.py -a ${ARCHIVE}/ -b build-android -H ${HOST} -s ${DEVICE}-m ${SOC} --checkpoint Llama3.2-1B-Instruct/consolidated.00.pth --params Llama3.2-1B-Instruct/params.json --tokenizer_model Llama3.2-1B-Instruct/tokenizer.model --prompt "<|start_header_id|>" --ptq 16a4w --temperature 0 --model_size 1B --seq_len 16  --model_mode kv

**Comment by cccclai** [comment]
> > Hey do you mind sharing the command for AoT and runtime so I can try on my end?
> 
> Sure! Change different mode (kv or bert )by set up the argument `model_mode`.
> 
> ```shell
> python examples/qualcomm/oss_scripts/llama3_2/llama.py -a ${ARCHIVE}/ -b build-android -H ${HOST} -s ${DEVICE}-m ${SOC} --checkpoint Llama3.2-1B-Instruct/consolidated.00.pth --params Llama3.2-1B-Instruct/params.json --tokenizer_model Llama3.2-1B-Instruct/tokenizer.model --prompt "<|start_header_id|>" --ptq 16a4w --temperature 0 --model_size 1B --seq_len 16  --model_mode bert
> ```

Ah I see - do you mind rename bert mode to batch_prefill? The context is that bert isn't a common name..

**Comment by chunit-quic** [comment]
> Ah I see - do you mind rename bert mode to batch_prefill? The context is that bert isn't a common name..

No problem. let me change it

**Comment by cccclai** [comment]
There are some errors here
```
executorch/examples/qualcomm/oss_scripts/llama3_2/runner/runner.cpp:50:7: error: field 'eval_mode_' will be initialized after field 'stats_' [-Werror,-Wreorder-ctor]
   50 |       eval_mode_(eval_mode),
      |       ^~~~~~~~~~~~~~~~~~~~~
      |       stats_({})
   51 |       stats_({}) {
      |       ~~~~~~~~~~
      |       eval_mode_(eval_mode)
```

**Comment by chunit-quic** [comment]
> executorch/examples/qualcomm/oss_scripts/llama3_2/runner/runner.cpp:50:7: error: field 'eval_mode_' will be initialized after field 'stats_' [-Werror,-Wreorder-ctor]
>    50 |       eval_mode_(eval_mode),
>       |       ^~~~~~~~~~~~~~~~~~~~~
>       |       stats_({})
>    51 |       stats_({}) {
>       |       ~~~~~~~~~~
>       |       eval_mode_(eval_mode)
> ```

Thanks for pointing out. Fixed.

File: install_requirements.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def python_is_compatible():
     "timm==1.0.7",
     f"torchaudio==2.5.0.{NIGHTLY_VERSION}" if USE_PYTORCH_NIGHTLY else "torchaudio",
     "torchsr==1.0.4",
-    "transformers==4.46.1",
+    "transformers==4.42.4", # TODO update back to 4.46.1 once the error is fixed
```
Comment (cccclai) [inline install_requirements.py:140]: What was the issue with this?

Comment (chunit-quic) [reply install_requirements.py]: Sorry for confusing. This is for our internal CI. Removed.

One more thing: just in case you want to reproduce the performance profiling results right now, we are still checking and working on some related passes. It might be better to wait for our next profiling results to see if the execution times are aligned. Thanks!

Comment (cccclai) [reply install_requirements.py]: Sounds good, thanks!

Comment (cccclai) [reply install_requirements.py]: Maybe let's bring more CI to oss, so they can be caught when creating the PR

**Comment by cccclai** [comment]
Still has lint error...

**Review by tarun292 — APPROVED** [review]

**Comment by cccclai** [comment]
`test-llama-runner-qnn-linux` seems failing...

**Comment by cccclai** [comment]
I'm getting following error:
```
I 00:00:00.000856 executorch:runner.cpp:54] creating module: model_path=llama3_2_qnn_batch_prefill.pte
I 00:00:00.000978 executorch:runner.cpp:56] creating runner: tokenizer_path=tokenizer.model
I 00:00:00.001265 executorch:runner.cpp:135] get_max_seq_len: 16
I 00:00:00.001331 executorch:runner.cpp:135] get_vocab_size: 128256
I 00:00:00.001370 executorch:runner.cpp:135] get_n_layers: 16
I 00:00:00.001419 executorch:runner.cpp:135] get_head_dim: 64
I 00:00:00.001476 executorch:runner.cpp:135] get_n_kv_heads: 8
I 00:00:00.443705 executorch:runner.cpp:81] creating io_memory
[INFO] [Qnn ExecuTorch]: create QNN Logger with log_level 2
[WARNING] [Qnn ExecuTorch]:  <W> Initializing HtpProvider

[WARNING] [Qnn ExecuTorch]:  <W> Function not called, PrepareLib isn't loaded!

[INFO] [Qnn ExecuTorch]: Initialize Qnn backend parameters for Qnn executorch backend type 2
[INFO] [Qnn ExecuTorch]: Caching: Caching is in RESTORE MODE.
[WARNING] [Qnn ExecuTorch]: Failed to interpret QNN context binary. Error code 30010. Try verifying binary with online-prepare format.
[ERROR] [Qnn ExecuTorch]: Failed to parse QNN Graph Info. The cache might be broken. Please consider to re-generate the cache.
[WARNING] [Qnn ExecuTorch]:  <W> Function not called, PrepareLib isn't loaded!

[ERROR] [Qnn ExecuTorch]: QNN context cache is invalid.
E 00:00:00.575785 executorch:QnnManager.cpp:303] Fail to configure Qnn context
E 00:00:00.575857 executorch:QnnExecuTorchBackend.cpp:60] Fail to initialize Qnn Manager
E 00:00:00.575926 executorch:method.cpp:109] Init failed for backend QnnBackend: 0x1
[WARNING] [Qnn ExecuTorch]:  <W> Backend 1 free cleanup called during process exit

[WARNING] [Qnn ExecuTorch]:  <W> qnnOpPackageManager: hexagon unload op package function pointer is nullptr!
```
The repro step is
1) AoT
```
python examples/qualcomm/oss_scripts/llama3_2/llama.py -a ./llama3_2_1B_512 -b build-android --compile_only -m "SM8650" --checkpoint /data/users/chenlai/models/llama3.2/1B/consolidated.00.pth --params /data/users/chenlai/models/llama3.2/1B/params.json --tokenizer_model /data/users/chenlai/models/llama3.2/1B/tokenizer.model --prompt "<|start_header_id|>" --ptq 16a4w --temperature 0 --model_size 1B --seq_len 16  --model_mode batch_prefill
```
2) Runtime, after pushing the binary/library to the device and set up the env path.
```
./qnn_llama3_2_runner --model_path llama3_2_qnn_batch_prefill.pte --tokenizer_path tokenizer.model                                                
```

**Comment by chunit-quic** [comment]
> `test-llama-runner-qnn-linux` seems failing, but it seems unrelated to this PR and it should be resolved in the main branch. Can you help with rebasing?

Sure, just rebased.

> I'm getting following error:
[WARNING] [Qnn ExecuTorch]: Failed to interpret QNN context binary. Error code 30010. Try verifying binary with online-prepare format.
[ERROR] [Qnn ExecuTorch]: Failed to parse QNN Graph Info. The cache might be broken. Please consider to re-generate the cache.
[WARNING] [Qnn ExecuTorch]:  <W> Function not called, PrepareLib isn't loaded!
[ERROR] [Qnn ExecuTorch]: QNN context cache is invalid.

May I know which QNN version did you use? It seems to me that it might relate to [PR6811](https://github.com/pytorch/executorch/pull/6811). I just tested with smaller one(1layer) without encountering an error. My qnn sdk version is 2.26.1

**Comment by cccclai** [comment]
I'm using the version downloaded from https://softwarecenter.qualcomm.com/api/download/software/qualcomm_neural_processing_sdk/v2.26.0.240828.zip 

Maybe let me get your latest pr and see if it passes. Seems like both batch_prefill and kv fails

**Comment by cccclai** [comment]
Is the PR tested on qnn 2.28?

**Comment by chunit-quic** [comment]
> Is the PR tested on qnn 2.28?

No, we test on qnn 2.26 previously. I'm running patch_prefill mode with 16 layers now with clean build again, and will do kv mode later.

**Comment by chunit-quic** [comment]
> No, we test on qnn 2.26 previously. I'm running patch_prefill mode with 16 layers now with clean build again, and will do kv mode later.

Just a quick update.

I run and succeed both kv and batch_prefill mode (16 layers, normal size) with the command below by changing the arg `--model_mode`. Didn't see the graph info error. My qnn sdk is 2.26.1.240827. 

If you would like to export and execute seperately, maybe you could consider using `--pre_gen_pte` arg of llama.py. It will help you to push libs and generate command on device.

Please let me know if you still facing the error. Would try to find it out. :)
```bash
 python examples/qualcomm/oss_scripts/llama3_2/llama.py -a ./${bert_test} -b build-android -H ${HOST} -s ${DEVICE} -m "SM8650" --checkpoint ${Llama3.2-1B-Instruct}/consolidated.00.pth --params ${Llama3.2-1B-Instruct}/params.json --tokenizer_model ${Llama3.2-1B-Instruct}/tokenizer.model --prompt "<|start_header_id|>" --ptq 16a4w --temperature 0 --model_size 1B --seq_len 16  --model_mode batch_prefill
```

**Comment by cccclai** [comment]
Sort of orthogonal - did we verify the runner cpp correctness? We have a calibrated quantized model which shows reasonable result, but the runner gives purely non-sense result.

**Comment by chunit-quic** [comment]
> Sort of orthogonal - did we verify the runner cpp correctness? We have a calibrated quantized model which shows reasonable result, but the runner gives purely non-sense result.

Yes, we performed some correctness checks.
1. Given a sequence of 15 tokens in prefill mode, we obtained its results (15 tokens). We then fed the same prompts to kv mode and obtained its results one by one (also 15 tokens in the end). The tokens from kv and prefill modes seem to be almost the same. (llama3_2 1b fp)
2. We have another internal PR based on this one to check correctness using stories 110M. The flow involves using prefill mode for prompts and kv mode to generate tokens based on the logits and kv cache from prefill mode. The results seem to be the same as pure kv mode. (16a4w)

We will update another PR based on this one, which supports hybrid mode. Maybe we can check correctness using that PR? Any other ideas are appreciated. :D

**Comment by cccclai** [comment]
Ah yes, that will be great. If we enable stories model, then we can add it to CI easily

**Comment by cccclai** [comment]
With that being said, should we just aim to merge that PR instead? and keep this PR on hold?

**Comment by chunit-quic** [comment]
> With that being said, should we just aim to merge that PR instead? and keep this PR on hold?

No, I would say we can merge this one first. Then the upcoming PR will have less code change.
